### PR TITLE
Refactor logins migration

### PR DESCRIPTION
**Pull request by skhamis**
Accidentally delete the previous branch https://github.com/mozilla/application-services/pull/4341 trying to fixup rebasing this ontop of logins-per-field-encryption so creating a new PR as I can't change the source of the previous one...

### Current Migration Plan
Step 1: We need to iterate through the old DB and throw it all into a struct that contains all the relevant information
```rust
struct MigrationPlan {
    logins: HashMap<String, MigrationLogin>,
}
```
```rust
struct MigrationLogin {
    local_login: Option<LocalLogin>,
    mirror_login: Option<MirrorLogin>,
    status: MigrationStatus,
}
```
Step 2: We need to perform options on the struct before throwing it into the new DB
 - Step 2a: Find any mismatched data between local/mirro and either throw away or attempt to reconcile
   - Case 1: If mirror has `is_override=1` but no matching local login
 - Step 2b: Find any logins we can fixup [LOCAL ONLY] and flip the `local_modified=true`

Step 3: Insert the struct data into the new DB
 - Step 3a: Capture the right metrics

Screen Shot 2021-08-30 at 1 55 07 PM

---

### Pull Request checklist ###

- [ ] **Quality**: This PR builds and tests run cleanly
  - Note:
    - For changes that need extra cross-platform testing, consider adding `[ci full]` to the PR title.
    - If this pull request includes a breaking change, consider [cutting a new release](https://github.com/mozilla/application-services/blob/main/docs/howtos/cut-a-new-release.md) after merging.
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes a changelog entry in [CHANGES_UNRELEASED.md](../CHANGES_UNRELEASED.md) or an explanation of why it does not need one
  - Any breaking changes to Swift or Kotlin binding APIs are noted explicitly
- [ ] **Dependencies**: This PR follows our [dependency management guidelines](https://github.com/mozilla/application-services/blob/main/docs/dependency-management.md)
  - Any new dependencies are accompanied by a summary of the due dilligence applied in selecting them.

### CODE REVIEW

File: components/logins/src/store.rs
Comment (lougeniaC64) [inline components/logins/src/store.rs:124]: We decided to keep the `add` and `update` functions separate in order to avoid the necessary UI updates so I think you don't need this stub.

File: components/logins/src/migrate_sqlcipher_db.rs
Comment (lougeniaC64) [inline components/logins/src/migrate_sqlcipher_db.rs:10]: nit: you can include `SyncStatus` in the import statement below

File: components/logins/src/migrate_sqlcipher_db.rs
Comment (lougeniaC64) [inline components/logins/src/migrate_sqlcipher_db.rs:85]: I'm wondering if the `migrate_mirror_logins` and `migrate_local_logins` functions should share the same transaction. I don't think we'd want either table to be migrated if the other isn't. IIRC in logins a combination of the records in the local and mirror table represent what was last synced so having them be in incongruous states seems undesirable. If they share this transaction, I would argue that `migrate_sync_metadata` should be included and metrics (with the exception of fixup metrics) should be collected for it.

**Comment by lougeniaC64** [comment]
I left a few comments in-line but I also want to raise the fact that I'm unclear on where we landed as far as what state we'll leave the database in if we can't successfully migrate a table. It may be that my thoughts on transactions aren't in line with that decision. Disregard if that's the case.

File: components/logins/src/migrate_sqlcipher_db.rs
Comment (skhamis) [reply components/logins/src/migrate_sqlcipher_db.rs:85]: Hmmm that does raise a few more questions too, we also probably need to revisit the discussion of end-to-end flow of rollbacks (rollforwards?) and everything inbetween:

1. I initially had it as separate tx just for ease of getting this code out, but definitely agree that maybe sharing the same transaction may help with the potential rollback logic.
  - However, it will be rare for an entire DB to be corrupted and most likely it'll be a few rows that will not be able to be fixed up/migrating at that point, do we also skip if we find that guid in the other table? From my understanding, this will only happen if they have a sync account but is migrating before syncing.
  - Since we're doing row by row instead of the entire table, if we cannot fixup a row, we just skip it -- lessening the reliance on needing to do any kind of major rollback (unless something goes horribly wrong with trying to fixup that row)
2. Since this is kind of a one-way street kind of migration. We most likely need to have a discussion of how we want to proceed if we can't even get the DB to open (I think our current stance is, trash it and start fresh) but as mentioned above, since we are doing row-by-row there is less chance of really bad things happening since we can just skip that row and move on.

Comment (lougeniaC64) [reply components/logins/src/migrate_sqlcipher_db.rs:85]: > rollbacks (rollforwards?) 

LOL, right. Rolling back probably isn't a viable option. We'd have to migrate the database away from SQLCipher at some point and what would change between the time the rollback occurred and the next migration attempt that would increase the likelihood of success? So, yes, rollforwards I think.

>     1. I initially had it as separate tx just for ease of getting this code out, but definitely agree that maybe sharing the same transaction may help with the potential rollback logic.

Yeah after I suggested the shared transaction approach, I started second-guessing it. If we're not rolling back, and I don't think we should, then we'd either be skipping records that can't be ported over or trashing databases in the corrupted db scenario. 

However I think we do need to ensure that the tables are in sync (🤡) with each other. If that's the case, we could remove records in the corresponding tables if they can't be migrated as you suggested. However what's to stop us from resetting sync and forcing a sync after the migration? That's a Friday EOD question but I'm not deleting it.

Comment (mhammond) [reply components/logins/src/migrate_sqlcipher_db.rs:116]: Yeah, I agree entirely with this - the problem doing local then the mirror is exactly that the tables might get out of sync. But the more I think about it, that *might* actually be an easily recoverable outcome: Consider:

* failing to migrate the `local` record but successfully migrating the `mirror` record: That means "local changes" are lost. Non sync users have lost this record entirely, Sync users will still work based on the mirror, but will have lost a local change waiting to sync up. That should be rare. I think the only problem here is that `is_overridden` would be wrong - so we should "migrate" that based on whether the actual record exists rather than the old DB value.

* successful `local` record but failing mirror record: should be impossible for non-sync users (they have no mirror records), and I think this would still work fine for sync users if we set `sync_status` to `SyncStatus::New/Changed`

A final consideration is: what happens when we've done a fixup? Does a fixup mean a "local change" (ie, that we should mark the item as locally modified and sync it up?) Fixing up a mirror record for a sync user almost certainly means a corrupt DB - that should exactly reflect the server - but we probably should not discard it entirely unless we have a 'local' record (see above, a "fixed up" local mirror record is better than discarding it because sync users, but we almost certainly want to arrange for us to re-download the mirror for that record)

I think the only way to capture all this (except metadata) as a single loop - ideally with a union statement, but if 2 read loops made it much easier that'd be ok. Then a struct/enum/handwave that represents the entire set of local states a GUID can be in and some rust pattern matching to deal with it. As as @lougeniaC64 mentions, a reasonable decision in some cases might be "also decline to migrate metadata, so we do a full server reconcile")

**Review by mhammond — COMMENTED** [review]
Thanks Sammy, this is looking great! Sorry it's been some time since I last looked in detail and while this is in great shape, I do think we should consider doing the migration as a single loop for the local/mirror records, as I tried to describe below :) We can chat more about this tomorrow.

**Comment by skhamis** [comment]
@bendk @mhammond as per conversation -- this is not meant to be the end-production migration but to get us 90% there. I will have a follow up document as well as update this PR issue to denote things to improve on in the future and edge cases to think about

**Review by mhammond — APPROVED** [review]
Thanks Sammy, this is in fairly good shape! It looks like the "happy case" is going to be happy, so I'm OK with landing this (although the test failures should be addressed before we do). However, I do think some of the error edge cases aren't quite going to work correctly, so there's still some stuff to do in followups - we can discuss how exactly to manage that later.

FWIW, I'm about to put a PR up to your branch, which might help explain some of the edge cases I describe here - that's really more to help explain what I said here - you can either suck that in before merging (it has `[ignore]` for the test skeletons) or after merging I'll update it to target the main branch.

File: components/logins/src/migrate_sqlcipher_db.rs
Comment (bendk) [inline components/logins/src/migrate_sqlcipher_db.rs:275]: This makes me think that the fixup should be a separate pass and `fixup_mismatched_records` should be last, so the code to handle the `is_overridden=true` but no local record is all in 1 place.  If needed we could collect some data in `MigrationLogin` to distinguish the different cases.  Maybe `local` wouldn't be an `Option`, it would be an enum with several cases:
  - `NotFound`
  - `Invalid`
  - `Valid(Login)`

But that's only if we actually need different behavior for those cases.

**Review by bendk — APPROVED** [review]
I like the direction of this a lot.  Seems like a few things need to be ironed out, and we need to decide how to handle some edge cases.  The main ones for me are:
  - We have an error writing to the new DB
  - `is_overridden=true` but there's no local login or a local login with invalid data.

So there's still some things to do, but I'm for merging and iterating on this one.

File: components/logins/src/migrate_sqlcipher_db.rs
Comment (skhamis) [reply components/logins/src/migrate_sqlcipher_db.rs:41]: This method was mainly for removing/adjusting things from the plan before we attempt to migrate all the records. I might've had the assumption that we would leave it up to the server to reconcile and flip the is_overridden as needed for all other cases (is_overridden but no local copy means the server cannot effectively reconcile, where the other cases might?)

**Comment by skhamis** [comment]
Thanks for the thorough review @mhammond & @bendk! I felt most of the suggested changes/tweaks were all valid and ended up removing a bunch of cruft to leave this feeling a little slicker. I ended up removing `fix_mismatched_records` as Ben was right, it doesn't really make sense to be it's own phase before we do all the fixups since we still need to do certain changes based on if we were able to import local vs mirror successfully. I think It'll be better served to rethink that approach and group that with the next set of changes (as far as I understand, we simply need a couple more checks + branches of code to complete this).

I'll probably go ahead and close this out and capture any remaining work in subsequent tasks to keep the PRs smaller.

**Pull request merged**
